src/components/Chat/hooks/useCreateChatClient.ts

```diff
@@ -0,0 +1,47 @@
+import { useEffect, useState } from 'react';
+
+import {
+  DefaultGenerics,
+  ExtendableGenerics,
+  OwnUserResponse,
+  StreamChat,
+  TokenOrProvider,
+  UserResponse,
+} from 'stream-chat';
+
+/**
+ * React hook to create, connect and return `StreamChat` client.
+ */
+export const useCreateChatClient = <SCG extends ExtendableGenerics = DefaultGenerics>({
+  apiKey,
+  tokenOrProvider,
+  userData,
+}: {
+  apiKey: string;
+  tokenOrProvider: TokenOrProvider;
+  userData: OwnUserResponse<SCG> | UserResponse<SCG>;
+}) => {
+  const [chatClient, setChatClient] = useState<StreamChat<SCG> | null>(null);
+
+  useEffect(() => {
+    const client = new StreamChat<SCG>(apiKey);
+    let didUserConnectInterrupt = false;
+
+    const connectionPromise = client.connectUser(userData, tokenOrProvider).then(() => {
+      if (!didUserConnectInterrupt) setChatClient(client);
+    });
+
+    return () => {
+      didUserConnectInterrupt = true;
+      setChatClient(null);
+      connectionPromise
+        .then(() => client.disconnectUser())
+        .then(() => {
+          console.log(`Connection for user "${userData.id}" has been closed`);
+        });
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [apiKey, userData.id, tokenOrProvider]);
```

Comment:
In case you want to handle this without ignoring `exhaustive-deps`:

```tsx
const [cachedUserData, setCachedUserData] = useState(userData);

if (userData.id !== cachedUserData.id) {
  setCachedUserData(userData);
}
```

And then you can safely add `cachedUserData` to the deps list.

Reply:
This adds extra overhead but it makes sense so I'll add it, thank you! 👍